File: apps/web/lib/core/models/gang_sheet.dart
/// DTF roll-width presets this print shop stocks, in millimeters — matches
/// the roadmap's "30 / 33 / 60 cm" plus a free-entry custom width.
const Map<String, int> gangSheetWidthPresetsMm = {
  "30 cm": 300,
  "33 cm": 330,
  "60 cm": 600,
};

enum GangSheetStatus {
  pending,
  ready,
  failed;

  static GangSheetStatus fromJson(String value) => GangSheetStatus.values
      .firstWhere((e) => e.name == value, orElse: () => GangSheetStatus.pending);

  bool get isTerminal => this == ready || this == failed;
}

/// One artwork placed on a gang sheet — both as requested (`copies`,
/// `rotateDeg`) and, once the sheet is ready, as actually laid out
/// (`x`/`y`/`width`/`height`, all millimeters).
class GangSheetItem {
  const GangSheetItem({
    required this.artworkId,
    this.copies = 1,
    this.rotateDeg = 0,
  });

  factory GangSheetItem.fromJson(Map<String, dynamic> json) => GangSheetItem(
        artworkId: json["artwork_id"] as String,
        copies: json["copies"] as int? ?? 1,
        rotateDeg: json["rotate_deg"] as int? ?? 0,
      );

  final String artworkId;
  final int copies;
  final int rotateDeg;

  Map<String, dynamic> toJson() =>
      {"artwork_id": artworkId, "copies": copies, "rotate_deg": rotateDeg};
}

class GangSheetPlacement {
  const GangSheetPlacement({
    required this.artworkId,
    required this.x,
    required this.y,
    required this.width,
    required this.height,
    required this.rotateDeg,
  });

  factory GangSheetPlacement.fromJson(Map<String, dynamic> json) =>
      GangSheetPlacement(
        artworkId: json["artwork_id"] as String,
        x: json["x"] as int,
        y: json["y"] as int,
        width: json["width"] as int,
        height: json["height"] as int,
        rotateDeg: json["rotate_deg"] as int,
      );

  final String artworkId;
  final int x;
  final int y;
  final int width;
  final int height;
  final int rotateDeg;
}

class GangSheet {
  const GangSheet({
    required this.id,
    required this.projectId,
    required this.status,
    required this.sheetWidthMm,
    required this.sheetHeightMm,
    required this.spacingMm,
    required this.dpi,
    required this.items,
    required this.layout,
    required this.errorMessage,
    required this.createdAt,
    this.downloadUrl,
  });

  factory GangSheet.fromJson(Map<String, dynamic> json) => GangSheet(
        id: json["id"] as String,
        projectId: json["project_id"] as String,
        status: GangSheetStatus.fromJson(json["status"] as String),
        sheetWidthMm: json["sheet_width_mm"] as int,
        sheetHeightMm: json["sheet_height_mm"] as int?,
        spacingMm: json["spacing_mm"] as int,
        dpi: json["dpi"] as int,
        items: (json["items"] as List<dynamic>? ?? [])
            .map((e) => GangSheetItem.fromJson(e as Map<String, dynamic>))
            .toList(),
        layout: (json["layout"] as List<dynamic>?)
            ?.map((e) => GangSheetPlacement.fromJson(e as Map<String, dynamic>))
            .toList(),
        errorMessage: json["error_message"] as String?,
        createdAt: DateTime.parse(json["created_at"] as String),
        downloadUrl: json["download_url"] as String?,
      );

  final String id;
  final String projectId;
  final GangSheetStatus status;
  final int sheetWidthMm;
  final int? sheetHeightMm;
  final int spacingMm;
  final int dpi;
  final List<GangSheetItem> items;
  final List<GangSheetPlacement>? layout;
  final String? errorMessage;
  final DateTime createdAt;
  final String? downloadUrl;
}

File: services/api/app/db/models/gang_sheet.py
from __future__ import annotations

import uuid
from typing import TYPE_CHECKING, Any

from sqlalchemy import Enum as SAEnum
from sqlalchemy import ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin, UUIDMixin
from app.db.enums import GangSheetStatus

if TYPE_CHECKING:
    from app.db.models.project import Project


class GangSheet(UUIDMixin, TimestampMixin, Base):
    """A multi-artwork auto-nested DTF print sheet, composited from the
    project's artworks onto one roll-width canvas.

    Lifecycle: PENDING -> READY | FAILED (mirrors Export).
    """

    __tablename__ = "gang_sheets"

    project_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    status: Mapped[GangSheetStatus] = mapped_column(
        SAEnum(GangSheetStatus, native_enum=False, length=16),
        default=GangSheetStatus.PENDING,
        nullable=False,
    )

    sheet_width_mm: Mapped[int] = mapped_column(Integer, nullable=False)
    # Computed by the auto-nest packer on success — the sheet's fixed width is
    # a roll-width preset, but its length grows to fit whatever was placed.
    sheet_height_mm: Mapped[int | None] = mapped_column(Integer)
    spacing_mm: Mapped[int] = mapped_column(Integer, nullable=False, default=5)
    dpi: Mapped[int] = mapped_column(Integer, nullable=False, default=300)

    # Requested items: [{artwork_id, copies, rotate_deg}].
    items: Mapped[list[dict[str, Any]]] = mapped_column(JSONB, nullable=False)
    # Computed on success: [{artwork_id, x, y, width, height, rotate_deg}] (mm).
    layout: Mapped[list[dict[str, Any]] | None] = mapped_column(JSONB)

    storage_bucket: Mapped[str | None] = mapped_column(String(120))
    storage_key: Mapped[str | None] = mapped_column(String(512))

    # Celery task id, used to reconcile worker state into this row on poll.
    task_id: Mapped[str | None] = mapped_column(String(155))
    error_message: Mapped[str | None] = mapped_column(Text)

    project: Mapped[Project] = relationship()
